ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp

```diff
@@ -191,80 +273,272 @@ static INT_PTR CALLBACK ShowMessageExtProc(HWND hwndDlg, UINT uMsg, WPARAM wPara
+  tstring tstrStr = widen(str);
 
-extern string window_get_caption();
+  wchar_t wstrWindowCaption[512];
+  GetWindowTextW(enigma::hWnd, wstrWindowCaption, 512);
 
-void message_alpha(double alpha) {
+  if (dialog_caption != L"")
+    wcsncpy_s(wstrWindowCaption, dialog_caption.c_str(), 512);
 
+  if (message_cancel) {
+    int result = MessageBoxW(enigma::hWnd, tstrStr.c_str(), wstrWindowCaption, MB_OKCANCEL | MB_ICONINFORMATION | MB_DEFBUTTON1 | MB_APPLMODAL);
+    if (result == IDOK) return 1; else return -1;
+  }
+
+  MessageBoxW(enigma::hWnd, tstrStr.c_str(), wstrWindowCaption, MB_OK | MB_ICONINFORMATION | MB_DEFBUTTON1 | MB_APPLMODAL);
+  return 1;
 }
 
-void message_background(int back) {
+static inline int show_question_helperfunc(string str) {
+  tstring tstrStr = widen(str);
 
-}
+  wchar_t wstrWindowCaption[512];
+  GetWindowTextW(enigma::hWnd, wstrWindowCaption, 512);
 
-void message_button(int spr) {
+  if (dialog_caption != L"")
+    wcsncpy_s(wstrWindowCaption, dialog_caption.c_str(), 512);
 
+  int result;
+  if (question_cancel) {
+    result = MessageBoxW(enigma::hWnd, tstrStr.c_str(), wstrWindowCaption, MB_YESNOCANCEL | MB_ICONQUESTION | MB_DEFBUTTON1 | MB_APPLMODAL);
+    if (result == IDYES) return 1; else if (result == IDNO) return 0; else return -1;
+  }
+
+  result = MessageBoxW(enigma::hWnd, tstrStr.c_str(), wstrWindowCaption, MB_YESNO | MB_ICONQUESTION | MB_DEFBUTTON1 | MB_APPLMODAL);
+  if (result == IDYES) return 1;
+  return 0;
 }
 
-void message_button_font(string name, int size, int color, int style) {
+static inline string get_open_filename_helper(string filter, string fname, string dir, string title) {
+  OPENFILENAMEW ofn;
 
-}
+  filter = filter.append("||");
+  fname = remove_slash(fname);
 
-void message_caption(bool show, string str) {
-  gs_cap = str;
-}
+  tstring tstr_filter = widen(filter);
+  replace(tstr_filter.begin(), tstr_filter.end(), '|', '\0');
+  tstring tstr_fname = widen(fname);
+  tstring tstr_dir = widen(dir);
+  tstring tstr_title = widen(title);
 
-void message_input_color(int col) {
+  wchar_t wstr_fname[MAX_PATH];
+  wcsncpy_s(wstr_fname, tstr_fname.c_str(), MAX_PATH);
 
-}
+  ZeroMemory(&ofn, sizeof(ofn));
+  ofn.lStructSize = sizeof(ofn);
+  ofn.hwndOwner = enigma::hWnd;
+  ofn.lpstrFile = wstr_fname;
+  ofn.nMaxFile = MAX_PATH;
+  ofn.lpstrFilter = tstr_filter.c_str();
+  ofn.nFilterIndex = 0;
+  ofn.lpstrTitle = tstr_title.c_str();
+  ofn.lpstrInitialDir = tstr_dir.c_str();
+  ofn.Flags = OFN_EXPLORER | OFN_FILEMUSTEXIST | OFN_HIDEREADONLY | OFN_NOCHANGEDIR;
 
-void message_input_font(string name, int size, int color, int style) {
+  if (GetOpenFileNameW(&ofn) != 0)
+    return shorten(wstr_fname);
 
+  return "";
 }
 
-void message_mouse_color(int col) {
+static inline string get_open_filenames_helper(string filter, string fname, string dir, string title) {
+  OPENFILENAMEW ofn;
 
-}
+  filter = string(filter).append("||");
```

Comment:
This block of code is identical to the one above, with exception of the `OFN_ALLOWMULTISELECT` flag. All other differences are purely syntactical.

Reply:
Ok, I think I finally fixed it.